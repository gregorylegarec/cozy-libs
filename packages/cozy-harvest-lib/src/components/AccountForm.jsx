import React, { PureComponent } from 'react'
import { Form, Field as FinalFormField } from 'react-final-form'
import PropTypes from 'react-proptypes'

import Button from 'cozy-ui/react/Button'
import { translate, extend } from 'cozy-ui/react/I18n'
import Field from 'cozy-ui/react/Field'

import Manifest from '../Manifest'

const legacyOptions = options =>
  options.map(option => ({ ...option, label: option.name }))

export class AccountField extends PureComponent {
  render() {
    const { name, t, type } = this.props
    const label = t(`fields.${name}.label`)
    const fieldProps = {
      ...this.props,
      className: 'u-m-0', // 0 margin
      fullwidth: true,
      label,
      size: 'medium'
    }
    switch (type) {
      case 'dropdown':
        return (
          <Field
            {...fieldProps}
            type="dropdown"
            options={legacyOptions(fieldProps.options)}
          />
        )
      case 'password':
        return <Field {...fieldProps} />
      default:
        return <Field {...fieldProps} type="text" />
    }
  }
}

export class AccountFields extends PureComponent {
  render() {
    const { manifestFields, t } = this.props

    // Ready to use named fields array
    const namedFields = Object.keys(manifestFields).map(fieldName => ({
      ...manifestFields[fieldName],
      name: fieldName
    }))

    return (
      <div>
        {namedFields.map((field, index) => (
          <FinalFormField key={index} name={field.name}>
            {({ input }) => (
              <AccountField label={field.name} {...field} {...input} t={t} />
            )}
          </FinalFormField>
        ))}
      </div>
    )
  }
}

export class AccountForm extends PureComponent {
  constructor(props, context) {
    super(props, context)
    const { lang, locales } = props
    if (locales && lang) {
      extend(locales[lang])
    }
  }

  render() {
    const { account, fields, t } = this.props
    const sanitizedFields = Manifest.sanitizeFields(fields)
    const initialValues = account ? account.auth : {}
    return (
      <Form
        initialValues={initialValues}
        // eslint-disable-next-line no-console
        onSubmit={v => console.log(v)}
        render={({ values }) => (
          <div>
            <AccountFields manifestFields={sanitizedFields} t={t} />
            <Button
              className="u-mt-2 u-mb-1-half"
              extension="full"
              label={t('accountForm.submit.label')}
              onclick={() => alert(JSON.stringify(values, 0, 2))}
            />
          </div>
        )}
      />
    )
  }
}

AccountForm.propTypes = {
  account: PropTypes.object,
  fields: PropTypes.object.isRequired,
  locales: PropTypes.object
}

export default translate()(AccountForm)
