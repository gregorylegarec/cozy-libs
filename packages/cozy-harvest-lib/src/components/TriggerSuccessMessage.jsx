import React, { Component } from 'react'
import PropTypes from 'react-proptypes'

import Button from 'cozy-ui/react/Button'
import { translate } from 'cozy-ui/react/I18n'

import connectingImg from '../assets/images/connecting.svg'

const IDLE = 'idle'
const DONE = 'done'

export class TriggerSuccessMessage extends Component {
  state = {
    status: IDLE
  }

  onClick = () => {
    const { status } = this.state
    if (status !== DONE) {
      const { onDone } = this.props
      onDone()
    }
    this.setState({ status: DONE })
  }

  render() {
    const { t } = this.props
    const { status } = this.state
    const isDone = status === DONE
    return (
      <div className="u-ta-center">
        <img src={connectingImg} />
        <h2>{t('triggerSuccessMessage.title')}</h2>
        <p className="u-ta-left">{t('triggerSuccessMessage.description')}</p>
        <Button
          disabled={isDone}
          label={t('triggerSuccessMessage.button.label')}
          onClick={this.onClick}
        />
      </div>
    )
  }
}

TriggerSuccessMessage.propTypes = {
  onDone: PropTypes.func.isRequired
}

export default translate()(TriggerSuccessMessage)
