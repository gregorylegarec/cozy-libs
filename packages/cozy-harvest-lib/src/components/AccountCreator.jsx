import React, { PureComponent } from 'react'
import PropTypes from 'react-proptypes'

import { withMutations } from 'cozy-client'

import AccountForm from './AccountForm'
import { accountsMutations } from '../connections/accounts'
import { prepareAccountData } from '../helpers/accounts'

/**
 * Encapsulates an AccountForm and create an account with resulting data.
 */
export class AccountCreator extends PureComponent {
  create = async data => {
    const { createAccount, konnector, onCreate, onCreateSuccess } = this.props
    onCreate()
    const account = await createAccount(prepareAccountData(konnector, data))
    onCreateSuccess(account)
  }

  render() {
    const { konnector, submitting } = this.props
    return (
      <AccountForm
        fields={konnector.fields}
        locales={konnector.locales}
        oauth={konnector.oauth}
        onSubmit={this.create}
        submitting={submitting}
      />
    )
  }
}

AccountCreator.propTypes = {
  createAccount: PropTypes.func.isRequired,
  konnector: PropTypes.object.isRequired,
  submitting: PropTypes.bool,
  onCreate: PropTypes.func.isRequired,
  onCreateSuccess: PropTypes.func.isRequired
}

export default withMutations(accountsMutations)(AccountCreator)
