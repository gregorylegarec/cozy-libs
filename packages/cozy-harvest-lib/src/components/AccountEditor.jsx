import React, { PureComponent } from 'react'
import PropTypes from 'react-proptypes'

import { withMutations } from 'cozy-client'

import AccountForm from './AccountForm'
import { accountsMutations } from '../connections/accounts'
import { hydrateAccount } from '../helpers/accounts'

/**
 * Encapsulates an AccocuntForm of an existing account, and allow to update it.
 */
export class AccountEditor extends PureComponent {
  update = async data => {
    const { account, onUpdate, onUpdateSuccess, updateAccount } = this.props
    onUpdate()
    const updatedAccount = await updateAccount(hydrateAccount(account, data))
    console.debug({ updatedAccount })
    onUpdateSuccess(updatedAccount)
  }

  render() {
    const { account, konnector, submitting } = this.props
    console.debug('AcconutEditor.render', { account })
    return (
      <AccountForm
        fields={konnector.fields}
        initialValues={account && (account.auth || account.oauth)}
        locales={konnector.locales}
        oauth={konnector.oauth}
        onSubmit={this.update}
        submitting={submitting}
      />
    )
  }
}

AccountEditor.propTypes = {
  account: PropTypes.object,
  createAccount: PropTypes.func.isRequired,
  konnector: PropTypes.object.isRequired,
  submitting: PropTypes.bool,
  onUpdate: PropTypes.func.isRequired,
  onUpdateSuccess: PropTypes.func.isRequired
}

export default withMutations(accountsMutations)(AccountEditor)
