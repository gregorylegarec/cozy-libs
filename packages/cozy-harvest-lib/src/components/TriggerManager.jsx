import React, { Component } from 'react'
import PropTypes from 'react-proptypes'

import { withMutations } from 'cozy-client'

import AccountCreator from './AccountCreator'
import AccountEditor from './AccountEditor'
import TriggerSuccessMessage from './TriggerSuccessMessage'
import { triggersMutations } from '../connections/triggers'
import {
  buildKonnectorCron,
  buildKonnectorTriggerAttributes
} from '../helpers/triggers'

const IDLE = 'IDLE'
const RUNNING = 'RUNNING'
const LOGGED = 'LOGGED'
const SUCCESS = 'SUCCESS'

/**
 * Deals with konnector configuration, i.e encapsulate account creation or
 * edition
 * @type {Component}
 */
export class TriggerManager extends Component {
  state = {
    status: IDLE
  }

  /**
   * Creation/update start handler
   * Set the status to RUNNING as soon as an account is being created
   * or updated.
   */
  handleAccountMutation = () =>
    this.setState({
      status: RUNNING
    })

  /**
   * Account creation success handler
   * @param  {Object}  account Created io.cozy.accounts document
   * @return {Object}          io.cozy.jobs document, runned with account data
   */
  handleTriggerSuccessMessage = async account => {
    const { createTrigger, konnector } = this.props

    const trigger = await createTrigger(
      buildKonnectorTriggerAttributes({
        konnector,
        account,
        cron: buildKonnectorCron(konnector)
      })
    )

    return await this.launch(trigger)
  }

  /**
   * Account update success handler
   * @param  {Object}  account Updated io.cozy.accounts document
   * @return {Object}          io.cozy.jobs document, runned with account data
   */
  handleAccountUpdateSuccess = async account => {
    this.setState({
      updatedAccount: account
    })

    const { trigger } = this.props
    return await this.launch(trigger)
  }

  /**
   * Launches a trigger
   * @param  {Object}  trigger io.cozy.triggers document
   * @return {Promise}         [description]
   */
  launch = async trigger => {
    const { launchTrigger, onLoginSuccess, waitForLoginSuccess } = this.props

    const job = await waitForLoginSuccess(await launchTrigger(trigger))

    if (['queued', 'running'].includes(job.state)) {
      onLoginSuccess(trigger)
    }

    this.setState({
      status: SUCCESS
    })
  }

  render() {
    const { account, konnector, onDone } = this.props
    const { status, updatedAccount } = this.state
    const succeed = [LOGGED, SUCCESS].includes(status)
    const submitting = status === RUNNING

    return account ? (
      <AccountEditor
        account={updatedAccount || account}
        konnector={konnector}
        onUpdate={this.handleAccountMutation}
        onUpdateSuccess={this.handleAccountUpdateSuccess}
        submitting={submitting}
      />
    ) : succeed ? (
      <TriggerSuccessMessage onDone={onDone} />
    ) : (
      <AccountCreator
        konnector={konnector}
        onCreate={this.handleAccountMutation}
        onCreateSuccess={this.handleTriggerSuccessMessage}
        submitting={submitting}
      />
    )
  }
}

TriggerManager.propTypes = {
  account: PropTypes.object,
  createTrigger: PropTypes.func.isRequired,
  konnector: PropTypes.object.isRequired,
  launchTrigger: PropTypes.func.isRequired,
  onDone: PropTypes.func.isRequired,
  onLoginSuccess: PropTypes.func.isRequired,
  trigger: PropTypes.object,
  waitForLoginSuccess: PropTypes.func.isRequired
}

export default withMutations(triggersMutations)(TriggerManager)
