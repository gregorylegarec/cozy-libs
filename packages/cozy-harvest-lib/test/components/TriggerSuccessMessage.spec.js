/* eslint-env jest */
import React from 'react'
import { configure, shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'

import { TriggerSuccessMessage } from 'components/TriggerSuccessMessage'

configure({ adapter: new Adapter() })

const t = jest.fn().mockImplementation(key => key)
const onDone = jest.fn()

describe('TriggerSuccessMessage', () => {
  beforeEach(() => {
    onDone.mockClear()
  })

  it('should render', () => {
    const component = shallow(
      <TriggerSuccessMessage onDone={onDone} t={t} />
    ).getElement()
    expect(component).toMatchSnapshot()
  })

  it('should call this.onClick on Button click', () => {
    const wrapper = shallow(<TriggerSuccessMessage onDone={onDone} t={t} />)
    wrapper.instance().onClick = jest.fn()
    wrapper.instance().forceUpdate()
    wrapper.find('DefaultButton').simulate('click')
    expect(wrapper.instance().onClick).toHaveBeenCalled()
  })

  it('should not call onDone twice', () => {
    const wrapper = shallow(<TriggerSuccessMessage onDone={onDone} t={t} />)
    wrapper.find('DefaultButton').simulate('click')
    wrapper.find('DefaultButton').simulate('click')
    expect(onDone).toHaveBeenCalledTimes(1)
  })

  it('should disable button after click', () => {
    const wrapper = shallow(<TriggerSuccessMessage onDone={onDone} t={t} />)
    wrapper.find('DefaultButton').simulate('click')
    const component = wrapper.getElement()
    expect(component).toMatchSnapshot()
  })
})
