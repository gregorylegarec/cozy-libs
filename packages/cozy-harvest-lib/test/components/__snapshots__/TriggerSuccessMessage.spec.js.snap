// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TriggerSuccessMessage should disable button after click 1`] = `
<div
  className="u-ta-center"
>
  <img
    src="file-mock"
  />
  <h2>
    TriggerSuccessMessage.title
  </h2>
  <p
    className="u-ta-left"
  >
    TriggerSuccessMessage.description
  </p>
  <DefaultButton
    disabled={true}
    label="TriggerSuccessMessage.button.label"
    onClick={[Function]}
  />
</div>
`;

exports[`TriggerSuccessMessage should render 1`] = `
<div
  className="u-ta-center"
>
  <img
    src="file-mock"
  />
  <h2>
    TriggerSuccessMessage.title
  </h2>
  <p
    className="u-ta-left"
  >
    TriggerSuccessMessage.description
  </p>
  <DefaultButton
    disabled={false}
    label="TriggerSuccessMessage.button.label"
    onClick={[Function]}
  />
</div>
`;
